packages/babel-parser/src/plugins/typescript/index.js

```diff
@@ -124,6 +127,24 @@ export default (superClass: Class<Parser>): Class<Parser> =>
       return undefined;
     }
 
+    tsParseModifiers<T: TsModifier>(allowedModifiers: T[]): { [*]: true } {
+      const modifiers = {};
+
+      while (true) {
+        const startPos = this.state.start;
+        const modifier = this.tsParseModifier(allowedModifiers);
+
+        if (!modifier) break;
+
+        if (Object.hasOwnProperty.call(modifiers, modifier)) {
+          this.raise(startPos, `Duplicate modifier: '${modifier}'`);
+        }
+        modifiers[modifier] = true;
+      }
```

Comment:
This seems to allow any modifier in any order; I presume that's okay? or are there some modifiers which have a specific required order (e.g. `declare` first)

Reply:
Some modifiers must come in a specific order, some others don't. For example, `abstract` and `readonly` can be swapped, but `public` must be before them.

We parse them using a logic similar to this one, which ensures the correct order:
```js
this.tsParseModifiers(["public"]);
this.tsParseModifiers(["abstract", "readonly"]);
```

Reply:
Not super important, but it might be nice to add a comment around modifier order/usage.